sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/FilterEstimation.scala

```diff
@@ -550,6 +565,225 @@ case class FilterEstimation(plan: Filter, catalystConf: CatalystConf) extends Lo
+        return None
+      case _ =>
+    }
+
+    val colStatLeft = colStatsMap(attrLeft)
+    val statsRangeLeft = Range(colStatLeft.min, colStatLeft.max, attrLeft.dataType)
+      .asInstanceOf[NumericRange]
+    val maxLeft = BigDecimal(statsRangeLeft.max)
+    val minLeft = BigDecimal(statsRangeLeft.min)
+
+    val colStatRight = colStatsMap(attrRight)
+    val statsRangeRight = Range(colStatRight.min, colStatRight.max, attrRight.dataType)
+      .asInstanceOf[NumericRange]
+    val maxRight = BigDecimal(statsRangeRight.max)
+    val minRight = BigDecimal(statsRangeRight.min)
+
+    // determine the overlapping degree between predicate range and column's range
+    val (noOverlap: Boolean, completeOverlap: Boolean) = op match {
+      // Left < Right or Left <= Right
+      // - no overlap:
+      //      minRight           maxRight     minLeft       maxLeft
+      // --------+------------------+------------+-------------+------->
+      // - complete overlap: (If null values exists, we set it to partial overlap.)
+      //      minLeft            maxLeft      minRight      maxRight
+      // --------+------------------+------------+-------------+------->
+      case _: LessThan =>
+        (minLeft >= maxRight,
+          maxLeft < minRight && colStatLeft.nullCount == 0 && colStatRight.nullCount == 0)
+      case _: LessThanOrEqual =>
+        (minLeft > maxRight,
+          maxLeft <= minRight && colStatLeft.nullCount == 0 && colStatRight.nullCount == 0)
+
+      // Left > Right or Left >= Right
+      // - no overlap:
+      //      minLeft            maxLeft      minRight      maxRight
+      // --------+------------------+------------+-------------+------->
+      // - complete overlap: (If null values exists, we set it to partial overlap.)
+      //      minRight           maxRight     minLeft       maxLeft
+      // --------+------------------+------------+-------------+------->
+      case _: GreaterThan =>
+        (maxLeft <= minRight,
+          minLeft > maxRight && colStatLeft.nullCount == 0 && colStatRight.nullCount == 0)
+      case _: GreaterThanOrEqual =>
+        (maxLeft < minRight,
+          minLeft >= maxRight && colStatLeft.nullCount == 0 && colStatRight.nullCount == 0)
+
+      // Left = Right or Left <=> Right
+      // - no overlap:
+      //      minLeft            maxLeft      minRight      maxRight
+      // --------+------------------+------------+-------------+------->
+      //      minRight           maxRight     minLeft       maxLeft
+      // --------+------------------+------------+-------------+------->
+      // - complete overlap:
+      //      minLeft            maxLeft
+      //      minRight           maxRight
+      // --------+------------------+------->
+      case _: EqualTo =>
+        ((maxLeft < minRight) || (maxRight < minLeft),
+          (minLeft == minRight) && (maxLeft == maxRight) && colStatLeft.nullCount == 0
```

Comment:
You said "we need one more condition: minLeft == minRight".  Note that this condition is already included.

@gatorsmile was suggesting "(minRight == maxRight) && (minLeft == minRight) && (maxLeft == maxRight)".  This implies all 4 values (minLeft, maxLeft, minRight, maxRight)  are equal.  This is not what I mean by complete overlap.  I initially defined "complete overlap" as "complete range overlap".  For example, we have a test case: test("cint = cint4").  If we use @gatorsmile's definition, then the case "cint = cint4" will become partial overlap with selectivity 0.33, which will under-estimate the selectivity.   In order to avoid out-of-memory error, I prefer over-estimating rather than under-estimating.